cpp/src/strings/capitalize.cu

```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/case.hpp>
+#include <cudf/utilities/error.hpp>
+#include <strings/char_types/is_flags.h>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+namespace { // anonym.
+  
+  //base class for probing string
+  //manipulation memory load requirements;
+  //
+  struct probe_base
+  {
+    using char_info = thrust::pair<uint32_t,detail::character_flags_table_type>;
+    
+    probe_base(column_device_view const d_column,
+               character_flags_table_type const* d_flags,
+               character_cases_table_type const* d_case_table):
+      d_column_(d_column),
+      d_flags_(d_flags),
+      d_case_table_(d_case_table)
+    {  
+    }
+
+    __host__ __device__
+    column_device_view const get_column(void) const
+    {
+      return d_column_;
+    }
+
+    __device__
+    char_info get_char_info(char_utf8 chr) const
+    {
+      uint32_t code_point = detail::utf8_to_codepoint(chr);
+      detail::character_flags_table_type flag = code_point <= 0x00FFFF ? d_flags_[code_point] : 0;
+      return char_info{code_point,flag};
+    }
+
+    __device__
+    char_utf8 convert_char(char_info const& info) const
+    {
+      return detail::codepoint_to_utf8(d_case_table_[info.first]);
+    }
+    
+  private:
+    column_device_view const d_column_;
+    character_flags_table_type const* d_flags_;
+    character_cases_table_type const* d_case_table_;
+  };
+
+  
+  //base class for executing string modification:
+  //
+  struct execute_base
```

Comment:
Could we combine the two base classes?
The `execute_base` just has two more members than `probe_base`
Perhaps the `d_offsets` and `d_chars` could be optional members
```
    capitialize_base(column_device_view const d_column,
                 character_flags_table_type const* d_flags,
                 character_cases_table_type const* d_case_table,
                 int32_t const* d_offsets = nullptr,
                 char* d_chars = nullptr ): ...
```

Reply:
Done.